econml/_rlearner.py

```diff
@@ -0,0 +1,330 @@
+                return self.model.predict(X)
+        np.random.seed(123)
+        X = np.random.normal(size=(1000, 3))
+        y = X[:, 0] + X[:, 1] + np.random.normal(0, 0.01, size=(1000,))
+        est = _RLearner(ModelFirst(LinearRegression()),
+                        ModelFirst(LinearRegression()),
+                        ModelFinal(),
+                        n_splits=2, discrete_treatment=False, random_state=None)
+        est.fit(y, X[:, 0], X=np.ones((X.shape[0], 1)), W=X[:, 1:])
+
+    >>> est.const_marginal_effect(np.ones((1,1)))
+    array([0.99963147])
+    >>> est.effect(np.ones((1,1)), T0=0, T1=10)
+    array([9.99631472])
+    >>> est.score(y, X[:, 0], X=np.ones((X.shape[0], 1)), W=X[:, 1:])
+    9.736380060274913e-05
+    >>> est.model_final.model
+    LinearRegression(copy_X=True, fit_intercept=False, n_jobs=None,
+         normalize=False)
+    >>> est.model_final.model.coef_
+    array([0.99963147])
+    >>> est.score_
+    9.826232040878233e-05
+    >>> [mdl._model for mdl in est.models_y]
+    [LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
+          normalize=False),
+     LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
+          normalize=False)]
+    >>> [mdl._model for mdl in est.models_t]
+    [LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
+          normalize=False),
+     LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
+          normalize=False)]
+
+    Attributes
+    ----------
+    models_y: list of objects of type(model_y)
+        A list of instances of the model_y object. Each element corresponds to a crossfitting
+        fold and is the model instance that was fitted for that training fold.
+    models_t: list of objects of type(model_t)
+        A list of instances of the model_t object. Each element corresponds to a crossfitting
+        fold and is the model instance that was fitted for that training fold.
+    model_final : object of type(model_final)
+        An instance of the model_final object that was fitted after calling fit.
+    score_ : float
+        The MSE in the final residual on residual regression, i.e.
+
+        .. math::
+            \\frac{1}{n} \\sum_{i=1}^n (Y_i - \\hat{E}[Y|X_i, W_i]\
+                                        - \\hat{\\theta}(X_i)\\cdot (T_i - \\hat{E}[T|X_i, W_i]))^2
+
+        If `sample_weight` is not None at fit time, then a weighted average is returned. If the outcome Y
+        is multidimensional, then the average of the MSEs for each dimension of Y is returned.
+    """
+
+    def __init__(self, model_y, model_t, model_final,
+                 discrete_treatment, n_splits, random_state):
+        class ModelNuisance:
+            """
+            Nuisance model fits the model_y and model_t at fit time and at predict time
+            calculates the residual Y and residual T based on the fitted models and returns
+            the residuals as two nuisance parameters.

```

Comment:
Perhaps a philosophical question, since I think the result is the same either way, but why not predict the actual Y and T (rather than calculate the residual) and perform the residualization within ModelFinal?  Wouldn't that more closely match the common description of the nuisances in this case?

Reply:
Yeah. seem the same. Thought this was neater, but no big deal. Residuals can also be equivalently thought as the nuisances.